Increase memory used for forked compiler

sciwhiz12 opened the pull request:
This PR fixes an out of memory error when compiling with a non-Java 21 JVM running the Gradle Daemon, by specifically setting the maximum memory used by the forked Java compiler process.

When running the Gradle Daemon with a non-J21 JVM, it will fork the Java compiler into a separate process rather than running it in-process with the Daemon. However, that means the compiler process uses the default amount of memory, which is not able to compile all of NeoForge & Minecraft (as it fails with an OutOfMemoryError).

Setting the memory when the compiler is forked ensures that whether the compiler runs in-process with the Daemon or as a forked process, it will always have enough memory to compile all the sources properly.

---

This is an alternative solution to the problem that #1692 also seeks to fix, which is more tailored to the problem by specifically addressing the root cause of the out-of-memory error.

To test, first set the Daemon to use a non-Java 21 JVM (via the `JAVA_HOME` environment variable pointed to a Java 17 JVM, for example), then run `./gradlew clean`, `./gradlew setup` (to prepare the workspace), then `./gradlew assemble --no-build-cache --rerun-tasks`. Without this PR, `:neoforge:compileJava` will fail with a "Java heap space" error; with this PR, the build will succeed.

sciwhiz12 committed:
Increase memory used for forked compiler

When running the Gradle Daemon with a non-J21 JVM, it will fork the
Java compiler into a separate process rather than running it in-process
with the Daemon. However, that means the compiler process uses the
default amount of memory, which is not able to compile all of NeoForge &
Minecraft (as it fails with an OutOfMemoryError).

Setting the memory when the compiler is forked ensures that whether the
compiler runs in-process with the Daemon or as a forked process, it will
always have enough memory to compile all the sources properly.

Technici4n commented on projects/neoforge/build.gradle:
```diff
@@ -222,6 +222,11 @@ generateAccessTransformers {
     )
 }
 
+tasks.withType(JavaCompile.class).configureEach {
```
Do we need the same in the base project?

sciwhiz12 replied:
Running with `./gradlew :base:compileJava :base:assemble --no-build-cache --rerun-tasks` with the same setup that causes the OOM error with `:neoforge:compileJava` does not cause an error, so it seems that the base subproject does not need the fix. That is also corroborated with the logs from my failed builds, which all run `:base:compileJava` successfully then fail on `:neoforge:compileJava`.

If wanted though as future-proofing, we could copy it into the base subproject or move it somewhere to be applied to both subprojects.

Technici4n replied:
Good enough as is then. If it happens to `base` in the future, we know how to fix it.

Technici4n commented on projects/neoforge/build.gradle:
```diff
@@ -222,6 +222,11 @@ generateAccessTransformers {
+    // Increase memory used during compilation, to avoid OutOfMemoryErrors
+    options.forkOptions.memoryMaximumSize = '2g'
+}
+
```
Should we also document the existence of this in `gradle.properties`, in case someone wants to change the Xmx? By the way, the `gradle.properties` can I think be changed back down to `2G`, and the comment can be updated as follows: `decompilation` -> `recompilation`.

sciwhiz12 replied:
Done.

sciwhiz12 committed:
Update memory comment in gradle.properties

sciwhiz12 committed:
Decrease Daemon JVM memory arg to 2G

The Daemon should require less memory as more things are now run out of
process. This also matches the max memory argument passed to the compile
tasks.

Matyrobbrt commented:
@neoforged/bots backport 1.20.6
